libraries/AP_Generator/AP_RichenPower.cpp

```diff
@@ -0,0 +1,430 @@
+{
+    return heat > heat_required_for_run();
+}
+
+constexpr float AP_RichenPower::heat_required_for_supply()
+{
+    // account for cooling that happens in that 60 seconds
+    return (30 * IDLE_RPM + 30 * RUN_RPM) * powf((1.0f-environment_loss_factor), 60);
+}
+bool AP_RichenPower::generator_ok_to_supply() const
+{
+    // duplicated into prearms
+    return heat > heat_required_for_supply();
+}
+
+
+
+/*
+  update the state of the sensor
+*/
+void AP_RichenPower::update(void)
+{
+    if (uart == nullptr) {
+        return;
+    }
+
+    update_servo_channel();
+
+    update_runstate();
+
+    if (get_reading()) {
+        // update
+        //printf("Data is %.2f\r\n",hybridData);
+    } else if (AP_HAL::millis() - last_reading_ms > 2000) {
+//        printf("No hybrid data\r\n");
+    }
+
+    update_heat();
+
+    Log_Write();
+}
+
+void AP_RichenPower::update_runstate()
+{
+    if (_servo_channel == nullptr) {
+        // doesn't really matter what we want to do; we're not going
+        // to be able to affect it...
+        return;
+    }
+
+    // don't run the generator while the safety is on:
+    // if (hal.util->safety_switch_state() == AP_HAL::Util::SAFETY_DISARMED) {
+    //     _servo_channel->set_output_pwm(SERVO_PWM_STOP);
+    //     return;
+    // }
+
+    static const uint16_t SERVO_PWM_STOP = 1200;
+    static const uint16_t SERVO_PWM_IDLE = 1500;
+    static const uint16_t SERVO_PWM_RUN = 1900;
+
+    switch (runstate) {
+    case RunState::STOP:
+        _servo_channel->set_output_pwm(SERVO_PWM_STOP);
+        break;
+    case RunState::IDLE:
+        _servo_channel->set_output_pwm(SERVO_PWM_IDLE);
+        break;
+    case RunState::RUN:
+        // we must have
+        if (!generator_ok_to_run()) {
+            _servo_channel->set_output_pwm(SERVO_PWM_IDLE);
+            break;
+        }
+        _servo_channel->set_output_pwm(SERVO_PWM_RUN);
+        break;
+    }
+}
+
+void AP_RichenPower::update_servo_channel(void)
+{
+    const uint32_t now = AP_HAL::millis();
+    if (now - _last_servo_channel_check < 1000) {
+        return;
+    }
+    _last_servo_channel_check = now;
+    SRV_Channel *control = SRV_Channels::get_channel_for(SRV_Channel::k_richenpower_control);
+    if (control == nullptr) {
+        if (_servo_channel != nullptr) {
+            gcs().send_text(MAV_SEVERITY_INFO, "RichenPower: no control channel");
+            _servo_channel = nullptr;
+        }
+        return;
+    }
+    if (_servo_channel != nullptr &&
+        _servo_channel->get_function() == SRV_Channel::k_richenpower_control) {
+        // note that _servo_channel could actually be == control here
+        return;
+    }
+    // gcs().send_text(MAV_SEVERITY_INFO, "RP: using control channel");
+    _servo_channel = control;
+}
+
+void AP_RichenPower::Log_Write()
+{
+    if (last_logged_reading_ms == last_reading_ms) {
+        return;
+    }
+    last_logged_reading_ms = last_reading_ms;
+
+    AP::logger().Write(
+        "RICH",
```

Comment:
I think we should perhaps use a more generic logging name.  Like "GEN" for generator.